test/e2e/tests/infra/infra.go

```diff
@@ -0,0 +1,336 @@
+
+	err := i.deployments.CreateOrUpdateAndWait(ctx, i.conf.AROResourceGroup, "ci-infra", mgmtfeatures.Deployment{
+		Properties: &mgmtfeatures.DeploymentProperties{
+			Template: t,
+			Mode:     mgmtfeatures.Incremental,
+		},
+	})
+
+	if azureerrors.IsDeploymentActiveError(err) {
+		err = i.deployments.Wait(ctx, i.conf.AROResourceGroup, "ci-infra")
+	}
+
+	if azureerrors.HasAuthorizationFailedError(err) ||
+		azureerrors.HasLinkedAuthorizationFailedError(err) {
+		return err
+	}
+
+	serviceErr, _ := err.(*azure.ServiceError) // futures return *azure.ServiceError directly
+
+	// CreateOrUpdate() returns a wrapped *azure.ServiceError
+	if detailedErr, ok := err.(autorest.DetailedError); ok {
+		serviceErr, _ = detailedErr.Original.(*azure.ServiceError)
+	}
+
+	if serviceErr != nil {
+		b, _ := json.Marshal(serviceErr)
+
+		return &api.CloudError{
+			StatusCode: http.StatusBadRequest,
+			CloudErrorBody: &api.CloudErrorBody{
+				Code:    api.CloudErrorCodeDeploymentFailed,
+				Message: "Deployment failed.",
+				Details: []api.CloudErrorBody{
+					{
+						Message: string(b),
+					},
+				},
+			},
+		}
+	}
+
+	return err
+}
+
+func (i *Infrastructure) deployResourceGroup(ctx context.Context) error {
+	_, err := i.rgClientSet.CreateOrUpdate(ctx, i.conf.AROResourceGroup+"-"+i.conf.RandSuffix, mgmtfeatures.ResourceGroup{Location: &i.conf.Location})
+	return err
+}
+
+// DeleteApp will be removed once the old apps are deleted
+// TODO: delete this exported function, no longer required
+func (i *Infrastructure) DeleteApp(ctx context.Context, appOID string) error {
+	_, err := i.appClientSet.Delete(ctx, appOID)
+	return err
+}
+
+func (i *Infrastructure) deployActiveDirectory(ctx context.Context) (string, string, string, string, error) {
+	t := time.Now().UTC().Truncate(time.Second)
+
+	appStartdate := date.Time{Time: t}
+	appEndDate := date.Time{Time: t.Add(48 * time.Hour)}
+	passID := uuid.NewV4()
+	pass := uuid.NewV4()
+
+	appName := i.conf.AROResourceGroup + "-" + i.conf.Cluster + i.conf.Location + "+" + i.conf.RandSuffix
+
+	app, err := i.appClientSet.Create(ctx, mgmtgraphrbac.ApplicationCreateParameters{
+		DisplayName:             to.StringPtr(appName),
+		Homepage:                to.StringPtr("https://" + appName),
+		IdentifierUris:          &[]string{"https://" + appName},
+		AvailableToOtherTenants: to.BoolPtr(false),
+		PasswordCredentials: &[]mgmtgraphrbac.PasswordCredential{
+			{
+				StartDate: &appStartdate,
+				EndDate:   &appEndDate,
+				KeyID:     to.StringPtr(passID.String()),
+				Value:     to.StringPtr(pass.String()),
+			},
+		},
+		ReplyUrls: &[]string{},
+	})
+
+	if err != nil {
+		return "", "", "", "", err
+	}
+
+	sp, err := i.spClientSet.Create(ctx, mgmtgraphrbac.ServicePrincipalCreateParameters{
+		AppID:          app.AppID,
+		AccountEnabled: to.BoolPtr(true),
+	})
+
+	if err != nil {
+		return "", "", "", "", err
+	}
+
+	return *app.AppID, *app.ObjectID, pass.String(), *sp.ObjectID, nil
```

Comment:
Why not return `app, sp, err` and parse outside?

Reply:
O think this might not be feasible for now :/

Reply:
I spent time on this, the pass was not returned as the object was created. And I did not want to burn more time on this.